### ISSUE TITLE
Sinatra 2.0.5 memory corruption when storing output of params[]

### ISSUE THREAD

**Issue by RootTJNII** [sinatra/sinatra#1549]
When storing the output of params[] in a downstream data structure the value is being corrupted across requests.

Take the following exemplar server:  https://github.com/RootTJNII/20190723SinatraBugExamplar/blob/master/server/app/api.rb

This is a minimal example app that has a single endpoint.
It caches the :id from the URL in a hash on the first request for :id, and returns the cached value on subsequent calls.

The following example client will trigger corruption in the hash: https://github.com/RootTJNII/20190723SinatraBugExamplar/blob/master/client.rb

This client generates 10 random hexadecimal IDs, then iterates over them in a loop until the server returns a 500.  Multiple :ids are needed to trigger the issue, calling only a single ID in a loop will not trigger the corruption.

When running the example client against the example server I'm seeing corruption in the stored values of the hash.  The server prints the cache to stdout on each call, and the values should only be `[0-9a-f]`.  Note the extra characters on the second run:

```
- -> /cache/f7643b320ba4077bfd4b682d2fd83092
     d627cdc1919aade634a83e3425bf7af7 => d627cdc1919aade634a83e3425bf7af7
     1c95795c08811648c6bdbc25545bbdcd => 1c95795c08811648c6bdbc25545bbdcd
     96f24b7bb4ee86c97ddf0433f8a712f9 => 96f24b7bb4ee86c97ddf0433f8a712f9
     7135b0fc16a9a1c8dd21d81aa51c1416 => 7135b0fc16a9a1c8dd21d81aa51c1416
     1748db142616624def07a0ea5f51d3ce => 1748db142616624def07a0ea5f51d3ce
     2aad574dfd5e88a32b3b18fed3d1b083 => 2aad574dfd5e88a32b3b18fed3d1b083
     c53f1d49a62c1f65adc6639af92cb615 => c53f1d49a62c1f65adc6639af92cb615
     604a0903fbc0a6a534ce8adc4b9fb0e1 => 604a0903fbc0a6a534ce8adc4b9fb0e1
     f7643b320ba4077bfd4b682d2fd83092 => f7643b320ba4077bfd4b682d2fd83092
     7933c4346a08cd96d45e146a4b0eabe7 => 7933c4346a08cd96d45e146a4b0eabe7
172.17.0.1 - - [23/Jul/2019:13:30:21 +0000] "GET /cache/7933c4346a08cd96d45e146a4b0eabe7 HTTP/1.1" 200 42 0.0050
172.17.0.1 - - [23/Jul/2019:13:30:21 UTC] "GET /cache/7933c4346a08cd96d45e146a4b0eabe7 HTTP/1.1" 200 42
- -> /cache/7933c4346a08cd96d45e146a4b0eabe7
     d627cdc1919aade634a83e3425bf7af7 => 627cdc1919aade634a83e342�
     1c95795c08811648c6bdbc25545bbdcd => c95795c08811648c6bdbc255p,
     96f24b7bb4ee86c97ddf0433f8a712f9 => 6f24b7bb4ee86c97ddf0433f Y
     7135b0fc16a9a1c8dd21d81aa51c1416 => 135b0fc16a9a1c8dd21d81aa

     1748db142616624def07a0ea5f51d3ce => 748db142616624def07a0ea5

     2aad574dfd5e88a32b3b18fed3d1b083 => aad574dfd5e88a32b3b18fed

     c53f1d49a62c1f65adc6639af92cb615 => 53f1d49a62c1f65adc6639af�
     604a0903fbc0a6a534ce8adc4b9fb0e1 => 04a0903fbc0a6a534ce8adc4�
     f7643b320ba4077bfd4b682d2fd83092 => 7643b320ba4077bfd4b682d2�
     7933c4346a08cd96d45e146a4b0eabe7 => 7933c4346a08cd96d45e146a4b0eabe7
2019-07-23 13:30:21 - JSON::GeneratorError - source sequence is illegal/malformed utf-8:
        /usr/local/lib/ruby/2.6.0/json/common.rb:224:in `generate'
        /usr/local/lib/ruby/2.6.0/json/common.rb:224:in `generate'
        /usr/local/lib/ruby/2.6.0/json/common.rb:394:in `dump'
        /app/api.rb:19:in `block in <top (required)>'
```

In testing with my full app where I originally spotted this bug I have seen NULLs and snippets of the HTTP headers where application values were expected.  My educated guess is there is a memory corruption issue in a lower module, it looks to me like the memory for params["id"] is being freed but not garbage collected properly, so the application hash is still holding a pointer to the free'd memory.

Basic instructions to reproduce are in https://github.com/RootTJNII/20190723SinatraBugExamplar.  Gemfile.lock for the exemplar is at https://github.com/RootTJNII/20190723SinatraBugExamplar/blob/master/server/app/Gemfile.lock.  I have reproduced this on multiple systems.

**Comment by RootTJNII** [comment]
Not able to reproduce using the Ruby:2.5 Docker container.  Now wondering if this is a Ruby bug and not a Sinatra bug.

**Comment by dentarg** [comment]
@RootTJNII just to be clear, you are able to reproduce this with Ruby 2.6? ([as specified in your Dockerfile](https://github.com/RootTJNII/20190723SinatraBugExamplar/blob/4d3fd54ed528864cd54d64a58ae9e32495c4be86/server/Dockerfile))

**Comment by RootTJNII** [comment]
@dentarg Correct.  I can reliably reproduce with the linked repo which pulls from the Ruby:2.6 container.

**Comment by dentarg** [comment]
What you are seeing is https://bugs.ruby-lang.org/issues/15792 and it has been fixed:  https://github.com/ruby/ruby/commit/3f9562015e651735bfc2fdd14e8f6963b673e22a, but there hasn't been any 2.6 release since. The fix is in `ruby 2.7.0preview1` and I'm not able to reproduce with 2.7.

**Comment by RootTJNII** [comment]
Great, thanks.  I was suspicious this was one of the recent use-after-close bugs I saw on bugs.ruby-lang.org, thank you for confirming.
